lib/rules/no-inferrable-types.js

```diff
@@ -0,0 +1,187 @@
+                        init.operator === "-" &&
+                        init.argument.type === "Identifier" &&
+                        init.argument.name === "Infinity") ||
+                    (init.type === "Identifier" && init.name === "Infinity")
+                ) {
+                    return true;
+                }
+
+                return (
+                    init.type === "Literal" && typeof init.value === "number"
+                );
+            }
+
+            return false;
+        }
+
+        /**
+         * Reports an inferrable type declaration, if any
+         * @param {ASTNode} node the node being visited
+         * @param {ASTNode} typeNode the type annotation node
+         * @param {ASTNode} initNode the initializer node
+         * @returns {void}
+         */
+        function reportInferrableType(node, typeNode, initNode) {
+            if (!typeNode || !initNode || !typeNode.typeAnnotation) {
+                return;
+            }
+
+            if (!isInferrable(typeNode, initNode)) {
+                return;
+            }
+
+            const typeMap = {
+                TSBooleanKeyword: "boolean",
+                TSNumberKeyword: "number",
+                TSStringKeyword: "string"
+            };
+
+            const type = typeMap[typeNode.typeAnnotation.type];
+
+            context.report({
+                node,
+                message: `Type ${type} trivially inferred from a ${type} literal, remove type annotation`,
+                fix: fixer => fixer.remove(typeNode)
+            });
+        }
+
+        /**
+         * Visits variables
+         * @param {ASTNode} node the node to be visited
+         * @returns {void}
+         */
+        function inferrableVariableVisitor(node) {
+            if (!node.id) {
+                return;
+            }
+            reportInferrableType(node, node.id.typeAnnotation, node.init);
+        }
+
+        /**
+         * Visits parameters
+         * @param {ASTNode} node the node to be visited
+         * @returns {void}
+         */
+        function inferrableParameterVisitor(node) {
+            if (ignoreParameters || !node.params) {
+                return;
+            }
+            node.params
+                .filter(
+                    param =>
+                        param.type === "AssignmentPattern" &&
+                        param.left &&
+                        param.right
+                )
+                .forEach(param => {
+                    reportInferrableType(
+                        param,
+                        param.left.typeAnnotation,
+                        param.right
+                    );
+                });
+        }
+
+        /**
+         * Visits properties
+         * @param {ASTNode} node the node to be visited
+         * @returns {void}
+         */
+        function inferrablePropertyVisitor(node) {
+            if (ignoreProperties || node.readonly) {
```

Comment:
How come `readonly` is being given special treatment here? It is not mentioned in your docs

Reply:
because of Microsoft/TypeScript#14416, this would not be typed as a boolean:

```ts
readonly foo = false;
```

for readonly properties, you need to specify the type explicitly. so this rule would end up suggesting invalid code if we didn't skip readonly properties.

i dont think its worth mentioning in the docs really as its to avoid introducing a bug we would've otherwise introduced. the expected behaviour is what we have

Reply:
Ah ok, please could you add a comment referencing that issue? It is really useful context for future readers of the code.